### ISSUE TITLE
pull mode: destination pool/dataset has been modified

### ISSUE THREAD

**Issue by eoli3n** [zrepl/zrepl#564]
### On A
```yaml
jobs:
- name: from_B
  type: pull
  connect:
    type: ssh+stdinserver
    host: server.fr
    user: root
    port: 22
    identity_file: /root/.ssh/id_rsa_zrepl
  root_fs: zroot/encrypted/zrepl
  interval: 10m
  pruning:
    keep_sender:
    - type: regex
      regex: ".*"
    keep_receiver:
    - type: grid
      grid: 1x1h(keep=all) | 24x1h | 30x1d | 6x30d | 1x365d
      regex: "^zrepl_"
```
### On B
```yaml
jobs:
- name: local
  type: snap
  filesystems: {
    "zroot<": true,
    "dpool/data<": true,
    "dpool/data/zrepl<": false
  }
  snapshotting:
    type: periodic
    prefix: zrepl_
    interval: 10m
  pruning:
    keep:
    - type: grid
      grid: 1x1h(keep=all) | 24x1h | 30x1d | 6x30d | 1x365d
      regex: "^zrepl_"

- name: to_A
  type: source
  snapshotting:
    type: manual
  serve:
    type: stdinserver
    client_identities:
      - "A-pull-from-B"
  filesystems: {
    "zroot<": true,
    "dpool/data<": false,
    "dpool/data/git<": true,
    "dpool/data/www<": true
  }
```

![pull](https://user-images.githubusercontent.com/14223657/149729129-4877bc80-fe97-4006-97e0-fdc49a445399.png)

The errors are related to the source side. ``zroot/bastille/jails/jekyll/root`` is on the ``source`` side.
Maybe that's because I use the ``snap`` workaround for pull mode ?
https://zrepl.github.io/configuration/prune.html#workaround-using-snap-job

> However, since, jobs are run independently, it is possible that the snap job will prune snapshots that are queued for replication / destruction by the remote pull job that connects to the source job. Symptoms of such race conditions are spurious replication and destroy errors.

In any cases, I do want to keep snapshotting in a separated task. Because of https://github.com/zrepl/zrepl/issues/559, I want to be able to disable ``source`` job, but not ``snap`` one.

**Comment by eoli3n** [comment]
I don't really get what "has been modified" ? If that's the data on the pool, not yet snapshotted, why does it matter ?
Why does zrepl check for this, as his work is to manage snapshots and to sync/prune them ?

**Comment by InsanePrawn** [comment]
> The errors are related to the source side. zroot/bastille/jails/jekyll/root is on the source side.

This error means that the dataset in question has been modified on the receiving(!) side. The complete sink dataset path is included. (`zroot/encrypted/zrepl/zroot/...`)

> I don't really get what "has been modified" ? If that's the data on the pool, not yet snapshotted, why does it matter ?
> Why does zrepl check for this, as his work is to manage snapshots and to sync/prune them ?

In order for ZFS to receive a new snapshot, the filesystem can't have been modified since the last snapshot. Changing metadata, such as atime, counts too. Another common cause is that zfs will create folder for auto-mounted child-datasets.

Tl;dr: Please set `canmount=off`, `readonly=on `, mountpoint=none` on your backup sink hierarchy.

In order to receive the snapshot, zrepl would have to force a rollback of the dataset to the last snapshot, which it will not do automatically due to the data loss involved.
There's #408 to track making that configurable, but that'd just be masking your symptoms.

**Comment by eoli3n** [comment]
Thanks
How to determine if a synced dataset will inherit or not of destination parent configurations ?
If I set ``mountpoint=none`` on ``zroot/encrypted/zrepl`` in which cases could ``zroot/encrypted/zrepl/zroot/ROOT/default`` still get a mountpoint ?
Should I ``inherit`` in ``recv`` options for all of those (atime, canmount, readonly, mountpoint) ?
https://zrepl.github.io/configuration/sendrecvoptions.html#job-recv-options-inherit-and-override

**Comment by eoli3n** [comment]
In the documentation, I don't find what tell what's the default behaviour. (inherit ?)

**Comment by eoli3n** [comment]
> In the documentation, I don't find what tell what's the default behaviour. (inherit ?)
Ok found https://zrepl.github.io/configuration/sendrecvoptions.html#job-note-property-replication

**Comment by problame** [comment]
@eoli3n please close this issue if you could resolve your issues.

@InsanePrawn Maybe we should categorically warn users if the receiving side doesn't have `canmount=off`, `readonly=on`, `mountpoint=none`? Which use cases would we break?

**Comment by Ramblurr** [comment]
This really bit me bad. I see that many other users have had this issue too.

Maybe at least the quickstart examples should be updated to include those options?

https://zrepl.github.io/quickstart/continuous_server_backup.html

**Comment by gdevenyi** [comment]
I also had this issue when I migrated from another replication solution
